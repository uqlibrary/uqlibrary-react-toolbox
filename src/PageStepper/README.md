# PageStepper

simple reusable page stepper component for UQ Library applications using ReactJs

## Props
- `searchResults: PropTypes.object`
- `children: PropTypes.array`
- `formSections: PropTypes.array`

## Usage

**./App.js**
```jsx
import React from 'react';
...

import {PageStepper} from 'uqlibrary-react-toolbox';
        
const App = () => (
    <PageStepper 
      formSections={['Page 1 Test', 'Page 2 Test']}
      >
        <AChildComponent data-stepperIndex="0" stepperHandleNext />
        <AnotherChildComponent data-stepperIndex="1" stepperHandlePrevious stepperOnsubmit />
    </PageStepper>
);

ReactDOM.render(
  <App />,
  document.getElementById('app')
);
```

