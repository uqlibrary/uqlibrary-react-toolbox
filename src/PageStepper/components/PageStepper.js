import React from 'react';
import ExpandTransition from 'material-ui/internal/ExpandTransition';
import PropTypes from 'prop-types';
import {Step, Stepper, StepLabel} from 'material-ui/Stepper';

class PageStepper extends React.Component {

    static propTypes = {
        searchResults: PropTypes.object,
        children: PropTypes.array,
        formSections: PropTypes.array
    };

    state = {
        loading: false,
        finished: false,
        stepIndex: 0
    };

    dummyAsync = (cb) => {
        this.setState({loading: true}, () => {
            this.asyncTimer = setTimeout(cb, 500);
        });
    };

    handleNext = () => {
        const {stepIndex} = this.state;

        if (!this.state.loading) {
            this.dummyAsync(() => {
                this.setState({
                    loading: false,
                    stepIndex: stepIndex + 1,
                    finished: stepIndex >= 2,
                });
            });
        }
    };

    handlePrev = () => {
        const {stepIndex} = this.state;
        if (!this.state.loading) {
            this.dummyAsync(() => this.setState({
                loading: false,
                stepIndex: stepIndex - 1,
            }));
        }
    };

    getComponent(children, step) {
        return children.filter( (comp) => {
            return comp.props['data-stepperIndex'] === step;
        });
    }

    // React.cloneElement allows us to add extra props onto a component dynamically
    updateProps() {
        return React.Children.map(this.props.children, child => {
            if (child) {
                return React.cloneElement(child, this.processProps(child));
            }

            return child;
        });
    }

    // optional functions that can be attached to a child component
    processProps(child) {
        let propsList = {};

        if (child) {
            if (child.props.stepperOnsubmit) {
                propsList = Object.assign({}, propsList, {onSubmit: this.handleNext});
            }

            if (child.props.stepperHandleNext) {
                propsList = Object.assign({}, propsList, {handleNext: this.handleNext});
            }

            if (child.props.stepperHandlePrevious) {
                propsList = Object.assign({}, propsList, {handlePrevious: this.handlePrev});
            }
        }

        return propsList;
    }

    render() {
        const {loading} = this.state;

        // update the props of each of the components to allow interaction with the stepper
        const children = this.updateProps();

        // retrieve all the components for the current stepIndex
        const currentComponent = this.getComponent(children, `${this.state.stepIndex}`);

        // build the step labels
        const steps = this.props.formSections.map((title, i) => {
            return (
                <Step key={i}>
                    <StepLabel style={{textOverflow: 'ellipsis', overflow: 'hidden'}}>{title}</StepLabel>
                </Step>
            );
        });

        return (
            <div className="layout-fill">
                <h1 className="page-title display-1">Add a journal article</h1>

                {/* Stepper start */}
                <Stepper activeStep={this.state.stepIndex} style={{padding: '0 25px', margin: '-10px auto' }} onChange={this.handleNext}>
                    {steps}
                </Stepper>


                <div style={{width: '100%', maxWidth: '1200px', margin: 'auto'}}>
                    <ExpandTransition loading={loading} open>
                        {currentComponent}
                    </ExpandTransition>
                </div>
                {/* Stepper end */}
            </div>
        );
    }
}

export default PageStepper;
