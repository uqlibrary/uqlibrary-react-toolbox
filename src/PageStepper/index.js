export {default as PageStepper} from './components/PageStepper';
