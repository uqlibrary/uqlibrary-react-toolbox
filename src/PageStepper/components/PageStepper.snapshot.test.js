jest.dontMock('./PageStepper');

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import PageStepper from './PageStepper';

function setup() {
    const props = {
        formSections: ['Page 1 Test', 'Page 2 Test']
    };

    return shallow(<PageStepper {...props}>
        <div data-stepperIndex="0">Page 1</div>
        <div data-stepperIndex="1">Page 2</div>
    </PageStepper>);
}

describe('PageStepper snapshots tests', () => {
    it('renders the stepper', () => {
        const app = setup();
        expect(toJson(app)).toMatchSnapshot();

        // go to page 2
        app.setState({stepIndex: 1});
        expect(toJson(app)).toMatchSnapshot();
    });

});


