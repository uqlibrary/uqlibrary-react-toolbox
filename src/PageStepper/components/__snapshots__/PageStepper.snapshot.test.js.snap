// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageStepper snapshots tests renders the stepper 1`] = `
<div
  className="layout-fill"
>
  <h1
    className="page-title display-1"
  >
    Add a journal article
  </h1>
  <Stepper
    activeStep={0}
    connector={<pure(StepConnector) />}
    linear={true}
    onChange={[Function]}
    orientation="horizontal"
    style={
      Object {
        "margin": "-10px auto",
        "padding": "0 25px",
      }
    }
  >
    <Step>
      <StepLabel
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Page 1 Test
      </StepLabel>
    </Step>
    <Step>
      <StepLabel
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Page 2 Test
      </StepLabel>
    </Step>
  </Stepper>
  <div
    style={
      Object {
        "margin": "auto",
        "maxWidth": "1200px",
        "width": "100%",
      }
    }
  >
    <ExpandTransition
      enterDelay={0}
      loading={false}
      open={true}
      transitionDelay={0}
      transitionDuration={450}
    >
      <div
        data-stepperIndex="0"
      >
        Page 1
      </div>
    </ExpandTransition>
  </div>
</div>
`;

exports[`PageStepper snapshots tests renders the stepper 2`] = `
<div
  className="layout-fill"
>
  <h1
    className="page-title display-1"
  >
    Add a journal article
  </h1>
  <Stepper
    activeStep={1}
    connector={<pure(StepConnector) />}
    linear={true}
    onChange={[Function]}
    orientation="horizontal"
    style={
      Object {
        "margin": "-10px auto",
        "padding": "0 25px",
      }
    }
  >
    <Step>
      <StepLabel
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Page 1 Test
      </StepLabel>
    </Step>
    <Step>
      <StepLabel
        style={
          Object {
            "overflow": "hidden",
            "textOverflow": "ellipsis",
          }
        }
      >
        Page 2 Test
      </StepLabel>
    </Step>
  </Stepper>
  <div
    style={
      Object {
        "margin": "auto",
        "maxWidth": "1200px",
        "width": "100%",
      }
    }
  >
    <ExpandTransition
      enterDelay={0}
      loading={false}
      open={true}
      transitionDelay={0}
      transitionDuration={450}
    >
      <div
        data-stepperIndex="1"
      >
        Page 2
      </div>
    </ExpandTransition>
  </div>
</div>
`;
